megamek/src/megamek/client/bot/princess/Princess.java

```diff
@@ -767,8 +767,13 @@ private Map<Mounted, Double> calcAmmoConservation(final Entity shooter) {
                     ammoCount += ammoCounts.get(ammoType);
                 }
                 msg.append(" has ").append(ammoCount).append(" shots left");
+                // Desired behavior:
```

Comment:
This can be tuned further by adding terms, but this should give us ammo weapon to-hit thresholds of:

- >= 12% (hit on 11+) when ammo is abundant; default Princess will take iffy shots but not if they require 12s.
- >= ~20% (hit on 9+) when ammo drops to 3 shots total; default Princess will try to close or hold fire at range.
- >= ~50% (hit on 7+) when ammo is down to 1 shot left.

Raising the aggroFactor (aggressiveness slider in Bot settings) or increasing ammo per weapon type lowers the computed thresholds.  Energy weapons are not affected.